source/Symbol/SwiftASTContext.cpp

```diff
@@ -4110,69 +4110,68 @@ swift::Type convertSILFunctionTypesToASTFunctionTypes(swift::Type t) {
 }
 
 CompilerType
-SwiftASTContext::GetTypeFromMangledTypename(const char *mangled_typename,
+SwiftASTContext::GetTypeFromMangledTypename(ConstString mangled_typename,
                                             Status &error) {
   VALID_OR_RETURN(CompilerType());
 
-  if (!mangled_typename ||
-      !SwiftLanguageRuntime::IsSwiftMangledName(mangled_typename)) {
+  const char *mangled_cstr = mangled_typename.AsCString();
+  if (mangled_typename.IsEmpty() ||
+      !SwiftLanguageRuntime::IsSwiftMangledName(mangled_cstr)) {
     error.SetErrorStringWithFormat(
-        "typename \"%s\" is not a valid Swift mangled name", mangled_typename);
+        "typename \"%s\" is not a valid Swift mangled name", mangled_cstr);
     return {};
   }
 
-  LOG_PRINTF(LIBLLDB_LOG_TYPES, "(\"%s\")", mangled_typename);
+  LOG_PRINTF(LIBLLDB_LOG_TYPES, "(\"%s\")", mangled_cstr);
 
   swift::ASTContext *ast_ctx = GetASTContext();
   if (!ast_ctx) {
     LOG_PRINTF(LIBLLDB_LOG_TYPES, "(\"%s\") -- null Swift AST Context",
-               mangled_typename);
+               mangled_cstr);
     error.SetErrorString("null Swift AST Context");
     return {};
   }
 
   error.Clear();
 
   // If we were to crash doing this, remember what type caused it.
-  llvm::PrettyStackTraceFormat PST("error finding type for %s",
-                                   mangled_typename);
-  ConstString mangled_name(mangled_typename);
-  swift::TypeBase *found_type =
-      m_mangled_name_to_type_map.lookup(mangled_name.GetCString());
+  llvm::PrettyStackTraceFormat PST("error finding type for %s", mangled_cstr);
+  swift::TypeBase *found_type = m_mangled_name_to_type_map.lookup(mangled_cstr);
   if (found_type) {
     LOG_PRINTF(LIBLLDB_LOG_TYPES, "(\"%s\") -- found in the positive cache",
-               mangled_typename);
+               mangled_cstr);
     assert(&found_type->getASTContext() == ast_ctx);
     return {found_type};
   }
 
-  if (m_negative_type_cache.Lookup(mangled_name.GetCString())) {
+  if (m_negative_type_cache.Lookup(mangled_cstr)) {
```

Comment:
We have a positive and a negative cache. Should we have a single cache instead that takes an `Optional<>` ?